### ISSUE TITLE
Anyone know why  half the lobby bots are MIA?

### ISSUE THREAD

**Issue by prastle** [triplea-game/triplea#4085]
### Engine version

NA
### My Operating System
NA
### Map name
NA

**Comment by prastle** [comment]
I had to stop and restart about 5 bots per server that were still running with stus check but MIA from the lobby.  Anyone know why? @RoiEXLab @ron-murhammer @DanVanAtta @ssoloff

**Comment by prastle** [comment]
Can someone also push a pr to infra so I can update the bots to at least this engine
https://github.com/triplea-game/triplea/releases/tag/1.9.0.0.12102
Thanks

**Comment by DanVanAtta** [comment]
> Can someone also push a pr to infra so I can update the bots to at least this engine

Actually, no. Our test server is dead. That means all deployments are frozen; we have not verification we are not deploying something totally broken. That verification should not be done on prod. The next steps are for us to fix prerelease server and then verify that the next version would be deployed well.

> I had to stop and restart about 5 bots per server that were still running with status check but MIA from the lobby. Anyone know why? @RoiEXLab @ron-murhammer @DanVanAtta @ssoloff

Which servers? 

@prastle side-note, feel free to replace the overview template with your first comment/question. It's not helpful to see the boiler plate questions we typically want asked left blank.

**Comment by prastle** [comment]
Then I guess we need to revert to a new stable that is lower
pick a number

**Comment by ron-murhammer** [comment]
@DanVanAtta Appears the bots are disappearing and not rejoining. There were like 5 of the 28 bots in the lobby this morning. I glanced at the logs quickly but didn't see any obvious cause. There are some errors around when people disconnect and not sure if those are new or always like that:
```
Sep 21 19:16:17 bot35_frankfurt_de run_bot.sh:  Sep 22, 2018 12:16:17 AM games.strategy.engine.framework.startup.launcher.ServerLauncher saveAndEndGame
Sep 21 19:16:17 bot35_frankfurt_de run_bot.sh:  INFO: Connection lost to:BoSSman23 game is over.  Game saved to:Bot304_FRANKFURT_DE_autosave.tsvg
Sep 21 19:16:17 bot35_frankfurt_de run_bot.sh:  Sep 22, 2018 12:16:17 AM games.strategy.engine.framework.startup.launcher.ServerLauncher lambda$launchInNewThread$2
Sep 21 19:16:17 bot35_frankfurt_de run_bot.sh:  SEVERE: Unrecognized error occurred: Exception on remote, if this is a repeatable error please make a copy of this savegame and report to:
Sep 21 19:16:17 bot35_frankfurt_de run_bot.sh:  https://github.com/triplea-game/triplea/issues/new
Sep 21 19:16:17 bot35_frankfurt_de run_bot.sh:  java.lang.RuntimeException: Exception on remote
Sep 21 19:16:17 bot35_frankfurt_de run_bot.sh:  #011at games.strategy.engine.message.UnifiedInvocationHandler.invoke(UnifiedInvocationHandler.java:54)
Sep 21 19:16:17 bot35_frankfurt_de run_bot.sh:  #011at games.strategy.engine.framework.$Proxy17.startPlayerStep(Unknown Source)
Sep 21 19:16:17 bot35_frankfurt_de run_bot.sh:  #011at games.strategy.engine.framework.ServerGame.waitForPlayerToFinishStep(ServerGame.java:535)
Sep 21 19:16:17 bot35_frankfurt_de run_bot.sh:  #011at games.strategy.engine.framework.ServerGame.runStep(ServerGame.java:423)
Sep 21 19:16:17 bot35_frankfurt_de run_bot.sh:  #011at games.strategy.engine.framework.ServerGame.startGame(ServerGame.java:290)
Sep 21 19:16:17 bot35_frankfurt_de run_bot.sh:  #011at games.strategy.engine.framework.startup.launcher.ServerLauncher.lambda$launchInNewThread$2(ServerLauncher.java:176)
Sep 21 19:16:17 bot35_frankfurt_de run_bot.sh:  #011at java.lang.Thread.run(Thread.java:748)
Sep 21 19:16:17 bot35_frankfurt_de run_bot.sh:  Caused by: games.strategy.engine.message.ConnectionLostException: Connection to BoSSman23 lost
Sep 21 19:16:17 bot35_frankfurt_de run_bot.sh:  #011at games.strategy.engine.message.UnifiedMessengerHub.connectionRemoved(UnifiedMessengerHub.java:176)
Sep 21 19:16:17 bot35_frankfurt_de run_bot.sh:  #011at games.strategy.net.AbstractServerMessenger.notifyConnectionsChanged(AbstractServerMessenger.java:493)
Sep 21 19:16:17 bot35_frankfurt_de run_bot.sh:  #011at games.strategy.net.AbstractServerMessenger.removeConnection(AbstractServerMessenger.java:655)
Sep 21 19:16:17 bot35_frankfurt_de run_bot.sh:  #011at games.strategy.net.AbstractServerMessenger.socketError(AbstractServerMessenger.java:672)
Sep 21 19:16:17 bot35_frankfurt_de run_bot.sh:  #011at games.strategy.net.nio.NioSocket.error(NioSocket.java:91)
Sep 21 19:16:17 bot35_frankfurt_de run_bot.sh:  #011at games.strategy.net.nio.NioReader.loop(NioReader.java:122)
Sep 21 19:16:17 bot35_frankfurt_de run_bot.sh:  #011... 1 more
```

**Comment by DanVanAtta** [comment]
`Caused by: games.strategy.engine.message.ConnectionLostException:` is a frequent error, a logging bug.

> But even if the bots crash, shouldn't they restart/rejoin? 

Yes. That would be  reasonable bug report. It will be difficult as we do not know how to repro, I think either way we'll be waiting for this problem to happen again.

Next time this happens we'll need to get thread dumps, those should tell us where the problem is.

**Comment by ssoloff** [comment]
> There are some errors around when people disconnect and not sure if those are new or always like that

#3672

**Comment by DanVanAtta** [comment]
Is this still a P1, all bots look connected. Am I missing any?

What action items do we want to come out of this? I suggest we spawn new issues to track those, I don't think there is much further we can do here.

I think one is to communicate and form a plan of how to get thread dumps before we restart. Any others?

**Comment by ron-murhammer** [comment]
I don't have any way to act on this but I think we should leave it open for at least a day or 2 to see if it occurs again.

**Comment by ssoloff** [comment]
Yeah, the focus should be on getting repro steps.  I don't know if simply letting a bot run in a local lobby on my box for > 24 hours is sufficient. :smile:

**Comment by DanVanAtta** [comment]
For how to handle next time to get more diagnostic info (thread dumps), I opened: https://github.com/triplea-game/triplea/issues/4094

IMO for this ticket:
- if a new incident happens, attaching those details to this ticket is not going to be helpful. This thread does not have much additional context to add, it'll be more to read. It's just as well to reference this issue to say this is suspect. If the new event turns out to be different, then we will not have mixed two different problems. Meanwhile, when we see a new event, a new issue is fast to create and if that gets very long, then we'll have helped to keep the issue trail short, which reduces our overhead when we get back up to speed on a problem.
- remove the P1 label, the impacts of this problem had been mitigated some time ago via manual restarts.
- identify any additional action items, create new issues for those, then close this one.

**Comment by ron-murhammer** [comment]
I'd say once #4094 is addressed then we can close this since we then have a plan for when this happens again.

**Comment by prastle** [comment]
this incident will happen again.  and ron was in the lobby when I joined.  A ghost?  that responded later?
either way yes I had to manually restart all of them.  roughly 20 in total on all severs.  I even asked Cernel if he knew when this had started.  He said it had been for a long time and thought I was doing something with them.  All comments on server thread and opening the issue are roughly thirty mins after I noticed and fixed.
